### DISCUSSION TITLE
How to export a ".mo" model as "FMU" in Linux?

### DISCUSSION THREAD

**Discussion by ejperezHI** [OpenModelica/OpenModelica#11531] [Q&A]
Which is the correct way of exporting .mo models as fmus in OpenModelica with Python or with the GUI itself?

I have 2 ways to do that in mind:

* Via the OpenModelica GUI
      With right click on the model>Export as>FMU
      This keeps giving me "Translation errors" everytime I try. What am I doing wrong? How is it done exactly?

* Via scripting
      Using "buildFMU" and so on commands, which appear here: https://build.openmodelica.org/Documentation/OpenModelica.Scripting.html
      However, everytime I tried to use them, gives me errors. How is it done exactly?

By the way, my OS is  Ubuntu 20.04.6 LTS and I am using OpenModelica 1.21.0~2-g0bc3d29

Thank you a lot in advance

**Comment by casella** [comment]
Please try OpenModelica 1.22.0. Just released. Lots of fixes.

**Comment by AnHeuermann** [comment]
Both ways are correct, see also the [User's Guide](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/fmitlm.html#fmi-export).

>This keeps giving me "Translation errors" everytime I try. What am I doing wrong? How is it done exactly?

I guess your model itself isn't correct and can't be translated into a simulation executable/FMU.
A FMU is nothing else than the usual simulation executable generated by the OpenModelica Compiler and than wrapped with FMI specific interface functions and zipped into one .fmu file.
Make sure that you can simulate your model before trying to export it as FMU.

Also note that you can only export a `model` not `packages` or whole .mo files.

**Reply by ejperezHI** [reply]
Hello, and thanks very much for the reply. 
I have tried simulating my model and it works fine. As said, I have tried those both methods nad noone works. 
Also, if I want to use those commands (https://build.openmodelica.org/Documentation/OpenModelica.Scripting.html), do I need to install a specific library/package/toolbox in Python?

**Reply by arun3688** [reply]
@ejperezHI can you just post your error message and we can see what is happening, or you can upload your model and we can see the problem

**Reply by ejperezHI** [reply]
Thank you a lot for your responses!
Here is the code I used for attempting to create a .fmu file out of a OpenModelica model:

`import openmodelica_microgrid_gym`
`from OMPython import OMCSession`
`import subprocess`
`import pyfmi`

`from OMPython import OMCSessionZMQ`
`omc = OMCSessionZMQ()`

`# Start session in OpenModelica`
`omc = OMCSessionZMQ()`

`# Charge the model`
`model_name = "EjemploOM_copia"`
`omc.sendExpression("loadModel(Modelica)")`
`omc.sendExpression(f"loadFile(\"{model_name}.mo\")")`

`# Export the model as .fmu file`
`export_command = f"buildModel({model_name}, outputFormat=\"fmu\")"`
`omc.sendExpression(export_command)`


And this is the result:
![image](https://github.com/OpenModelica/OpenModelica/assets/131266350/b5565f2d-d0db-462f-971b-4b5a438355b8)

Only that. Nothing happens or any .fmu file is created

Thanks,

**Reply by arun3688** [reply]
@ejperezHI you should use `buildModelFMU()` API and the code should be something like below

```
from OMPython import OMCSessionZMQ
omc = OMCSessionZMQ()
omc.sendExpression("loadModel(Modelica)")
omc.sendExpression("loadFile("\EjemploOM_copia.mo\")")
omc.sendExpression("buildModelFMU(EjemploOM_copia)")
omc.sendExpression("getErrorString()")

```

**Reply by arun3688** [reply]
@ejperezHI if you don't see the desired results you should use `getErrorString()` API to get the error message

**Reply by ejperezHI** [reply]
Thanks for your reply!
Done, this is the error that appears:

'Error: Failed to load package EjemploOM_copia (default) using MODELICAPATH /home/ejperez/.openmodelica/libraries/.\n[/var/lib/jenkins2/ws/LINUX_BUILDS/tmp.build/openmodelica-1.22.0/OMCompiler/Compiler/Util/System.mo:1091:5-1091:72:writable] Error: Internal error System.realpath failed\n'


Thanks,

**Reply by arun3688** [reply]
@ejperezHI i don't know how your package is structured, but i can also see some strange internal error i guess you have the same issue in omedit, if possible attach your model and let me see if i can reproduce the issue

**Reply by ejperezHI** [reply]
Thanks again for your reply.
Yes, it is the same error that appears in OMEdit. I just have installed OpenModelica 1.22 by the way...however, back the past week, even with OpenModelica 1.21, I have the same results as now exporting FMUs both via OMEdit and via scripting.

How do I attach the model .mo? It gives me an error saying that "We dn't support that fle type".

Thanks,

**Reply by arun3688** [reply]
@ejperezHI  attach as a zip file

**Reply by ejperezHI** [reply]
Here it is
[EjemploOM_copia.zip](https://github.com/OpenModelica/OpenModelica/files/13334971/EjemploOM_copia.zip)

Thanks!

**Comment by arun3688** [comment]
@ejperezHI  I tested your model and it works fine on my computer in OMEdit and also from OMPython, I can able to simulate the model and also export the model as FMU directly in OMEdit, see the image below

![image](https://github.com/OpenModelica/OpenModelica/assets/6870949/0deacd29-e264-4822-8c91-ee62ec9c5c50)

So i suspect there is some other issue going on with your computer or your omc version,  related to the issue you posted https://github.com/OpenModelica/OpenModelica/discussions/11561, I suggest you to completely uninstall OpenModelica and try to install the latest nightly builds version

**Reply by ejperezHI** [reply]
Thank you a lot for your helpful reply,

Yes, I know the model is fine and probably in windows it works perfectly as it is the operating system in which the model has been prepared to be exported. The issue is that my computer has Ubuntu 20.0.4 LTS and this is probably where the issue is. Last week I had the previous version of OpenModelica (the 1.21) and today I have uninstalled it (I think well) and installed the new version (1.22) according to the recommendation of the colleague @casella. However, I couldn't export to .fmu neither with the previous version (1.21) nor with this one (1.22).

Also, I have seen different kinds of the same versions of open modelica: the stable one, the release one and the nightly one. Which are the differences and why should one work better than other?

Thanks,

**Reply by arun3688** [reply]
@ejperezHI the nightly builds are the latest ones which are built every day and has lot of bug fixes, but i guess there is some ubuntu packages you might be missing i guess when installing openmodelica i guess, I don't have linux so i don't know what is the problem

**Reply by ejperezHI** [reply]
I just solved the issue posted in #11561 . I finally have the OpenModelica 1.22 perfectly installed in my Ubuntu computer.
And just as ocurred the past week when I had OpenModelica 1.22, the same happens.

Just as I said in the first post:

- When trying to export as FMU via OpenModelica GUI, this happens:
![image](https://github.com/OpenModelica/OpenModelica/assets/131266350/6a43fdf1-f109-4d49-b46b-d880789ce3aa)


- When trying to export as FMU via code:

- With the code written above:
`import openmodelica_microgrid_gym`
`from OMPython import OMCSession`
`import subprocess`
`import pyfmi`

`from OMPython import OMCSessionZMQ`
`omc = OMCSessionZMQ()`

`# Start session in OpenModelica`
`omc = OMCSessionZMQ()`

`# Charge the model`
`model_name = "EjemploOM_copia"`
`omc.sendExpression("loadModel(Modelica)")`
`omc.sendExpression(f"loadFile(\"{model_name}.mo\")")`

`# Export the model as .fmu file`
`export_command = f"buildModel({model_name}, outputFormat=\"fmu\")"`
`omc.sendExpression(export_command)`

the same happens, just this output:
![image](https://github.com/OpenModelica/OpenModelica/assets/131266350/dffb9242-2b0d-48e3-b62c-14d66fbaec53)

- And with this new code, this appears:
`try:`
    `omc = OMCSession()`

    `model_name = "EjemploOM_copia"`

    `omc.sendExpression("loadModel(Modelica)")`
    `omc.sendExpression(f"loadFile(\"{model_name}.mo\")")`

    `export_command = f"buildModel({model_name}, outputFormat=\"fmu\")"`
    `result = omc.sendExpression(export_command)`

    `print(result)`

`except Exception as e:`
    `print("Ocurrió un error:", e)`

![image](https://github.com/OpenModelica/OpenModelica/assets/131266350/f55054f4-fbef-4102-ba00-e773161aef22)


Thanks in advance!

**Reply by arun3688** [reply]
@ejperezHI the API you are using is wrong and you have to use `buildModelFMU`, You are trying to mix OMCSession and OMCSessionZMQ(), 

```
>>> from OMPython import OMCSessionZMQ
>>> omc = OMCSessionZMQ()
>>> print(omc.sendExpression("getVersion()"))
>>> omc.sendExpression("loadModel(Modelica)")
>>> omc.sendExpression("loadFile(\"EjemploOM_copia\")")
>>> omc.sendExpression("buildModelFMU(EjemploOM_copia)")
>>> omc.sendExpression("getErrorString()")
```

**Reply by ejperezHI** [reply]
Thanks for the reply,
I just tried that and this is the output:
![image](https://github.com/OpenModelica/OpenModelica/assets/131266350/e44b9cc4-7792-4f19-8be1-777ad4c62847)


Thanks,

**Reply by AnHeuermann** [reply]
For the CMake error in OMEdit: Simply update CMake using apt.
You can also disable the feature that relies on a recent CMake altogether by adding [`--fmuRuntimeDepends=none`](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/omchelptext.html#omcflag-fmuruntimedepends) to your translation flags. In OMEdit: Tools->Options->Simulation->Additional Translation Flags.

**Reply by ejperezHI** [reply]
Thanks for your reply!
Done. In the GUI, the same Translation errors as above appears, and with the code:

`from OMPython import OMCSessionZMQ`
 `omc = OMCSessionZMQ()`
 `print(omc.sendExpression("getVersion()"))`
 `omc.sendExpression("loadModel(Modelica)")`
 `omc.sendExpression("loadFile(\"EjemploOM_copia\")")`
 `omc.sendExpression("buildModelFMU(EjemploOM_copia, flags=\"--fmuRuntimeDepends=none\")")`
 `omc.sendExpression("getErrorString()")`

This is now what appears:
![image](https://github.com/OpenModelica/OpenModelica/assets/131266350/14077b07-7163-4d0e-9ad1-8758c2cd07f6)


Thanks,

**Reply by AnHeuermann** [reply]
Please read the [user's guide on the scripting API functions](https://openmodelica.org/doc/OpenModelicaUsersGuide/1.22/scripting_api.html#scripting-api) if you want to use them from OMPython. Function [`buildModelFMU`](https://openmodelica.org/doc/OpenModelicaUsersGuide/1.22/scripting_api.html#buildmodelfmu) has no input argument named `flags`.

Use [`setCommandLineOptions`](https://openmodelica.org/doc/OpenModelicaUsersGuide/1.22/scripting_api.html#setcommandlineoptions) to set compiler flags:

```python
from OMPython import OMCSessionZMQ
omc = OMCSessionZMQ()
print(omc.sendExpression("getVersion()"))
omc.sendExpression("loadModel(Modelica)")
omc.sendExpression("getErrorString()")                  # Add more getErrorString to find your issue
omc.sendExpression("loadFile(\"EjemploOM_copia.mo\")")  # Shouldn't it be \"EjemploOM_copia.mo\"?
omc.sendExpression("getErrorString()")
omc.sendExpression("setCommandLineOptions(\"--fmuRuntimeDepends=none\")")
omc.sendExpression("getErrorString()")
omc.sendExpression("buildModelFMU(EjemploOM)")          # Your model isn't named EjemploOM_copia but EjemploOM, at least the one you uploaded
omc.sendExpression("getErrorString()")
```

You'll get a warning, but the FMU export should be working.
```
'Warning: It was not possible to determine if the initialization problem is consistent, because of not evaluable parameters/start values during compile time: inductor.i = $START.inductor.i (sineCurrent.signalSource.offset + (if time < sineCurrent.signalSource.startTime then 0.0 else sineCurrent.signalSource.amplitude * sin(6.283185307179586 * sineCurrent.signalSource.f * (time - sineCurrent.signalSource.startTime) + sineCurrent.signalSource.phase)) = $START.inductor.i)\nWarning: The initial conditions are over specified. For more information set -d=initialization. In OMEdit Tools->Options->Simulation->Show additional information from the initialization process, in OMNotebook call setCommandLineOptions("-d=initialization").\n'
```

**Reply by ejperezHI** [reply]
Big thanks for the reply!
Done, but still doesn't work. This is my output after running it: 
The same error as above appears:
![image](https://github.com/OpenModelica/OpenModelica/assets/131266350/e44b9cc4-7792-4f19-8be1-777ad4c62847)

By the way, both "EjemploOM_copia" and "EjemploOM" are the exact same models in OpenModelica, just with different names, sprry for the misunderstanding.
Thanks,

**Reply by AnHeuermann** [reply]
Are you sure you fixed your `loadFile` and `buildModelFMU` commands? Both had errors in them which I fixed in my previous comment.
> omc.sendExpression("loadFile(\"EjemploOM_copia.mo\")")  # Shouldn't it be \"EjemploOM_copia.mo\"?

> omc.sendExpression("buildModelFMU(EjemploOM)")          # Your model isn't named EjemploOM_copia but EjemploOM, at least the one you uploaded

Make sure that `loadFile` returns `true` and that the error string is empty.

The error message tells you, that OpenModelica doesn't know where to find Modelica class `EjemploOM_copia`.

**Reply by ejperezHI** [reply]
Thanks again for your reply,
>Are you sure you fixed your loadFile and buildModelFMU commands? Both had errors in them which I fixed in my previous comment.

Yes, I writed exactly the code you gave me. 
>Make sure that loadFile returns true

Also, after checking it, I realized that the "loadFile" commands doesn't return a "True" but a "False". Why is that and how can I solve it? I wrote exactly what you gave me and it is okay according to the documentation.

>The error message tells you, that OpenModelica doesn't know where to find Modelica class EjemploOM_copia.

And how can I solve this? The OpenModelica model "EjemploOM.mo" and the python script is in the same directory. What should I do, include every "OpenModelica" model that I want to convert to FMU sinde the .openmodelica/libraries directory?

Thanks,

**Reply by AnHeuermann** [reply]
> Also, after checking it, I realized that the "loadFile" commands doesn't return a "True" but a "False". Why is that and how can I solve it? I wrote exactly what you gave me and it is okay according to the documentation.

Because the function failed. Check the error string to see why. I suspect a typo or a missing file extension `.mo` in the function call.
Maybe post your Python script again. 

> And how can I solve this? The OpenModelica model "EjemploOM.mo" and the python script is in the same directory. What should I do, include every "OpenModelica" model that I want to convert to FMU sinde the .openmodelica/libraries directory?

You'll need to find out why `loadFile` is failing.

**Reply by ejperezHI** [reply]
Thanks for the reply!

This is the code:

`from OMPython import OMCSessionZMQ`
`omc = OMCSessionZMQ()`
`print(omc.sendExpression("getVersion()"))`
`omc.sendExpression("loadModel(Modelica)")`
`omc.sendExpression("getErrorString()")`                  
`omc.sendExpression("loadFile(\"EjemploOM_copia.mo\")")`  
`omc.sendExpression("getErrorString()")`
`omc.sendExpression("setCommandLineOptions(\"--fmuRuntimeDepends=none\")")`
`omc.sendExpression("getErrorString()")`
`omc.sendExpression("buildModelFMU(EjemploOM)")`          
`omc.sendExpression("getErrorString()")`

How do I check the error more than the output itself that says what I have posted above? 

Thanks,

**Reply by AnHeuermann** [reply]
Print the strings returned by `getErrorString`.

**Reply by ejperezHI** [reply]
Thanks again for your reply!

Executing this:
`from OMPython import OMCSessionZMQ`
`omc = OMCSessionZMQ()`
`print(omc.sendExpression("getVersion()"))`
`omc.sendExpression("loadModel(Modelica)")`
`omc.sendExpression("getErrorString()")`                  
`omc.sendExpression("loadFile(\"EjemploOM_copia.mo\")")`  

I get this output (obviously a False, just as I commented in the previous post):
![image](https://github.com/OpenModelica/OpenModelica/assets/131266350/54af9971-74db-4f07-beed-9e91a7f3f752)

Executing this (with the getErrorString():
`from OMPython import OMCSessionZMQ`
`omc = OMCSessionZMQ()`
`print(omc.sendExpression("getVersion()"))`
`omc.sendExpression("loadModel(Modelica)")`
`omc.sendExpression("getErrorString()")`                  
`omc.sendExpression("loadFile(\"EjemploOM_copia.mo\")")`  
`omc.sendExpression("getErrorString()")`
I get this
![image](https://github.com/OpenModelica/OpenModelica/assets/131266350/a52cdf4f-15bb-48fe-a0d6-a3448b8b3baf)



Thanks,

**Reply by arun3688** [reply]
@ejperezHI can you write the `fullpath` for your file location for example
```

omc.sendExpression("loadFile(\"home/test/Ejemple.mo\")")
omc.sendExpression("getErrorString()")
```

**Reply by ejperezHI** [reply]
Thanks again for the quick reply,

Done, I get the exact same output
![image](https://github.com/OpenModelica/OpenModelica/assets/131266350/a52cdf4f-15bb-48fe-a0d6-a3448b8b3baf)

Thanks,

**Reply by arun3688** [reply]
@ejperezHI can you show the script again with full path, If it returns `false` means then you are providing the wrong path of the file location which you have to figure that out

**Reply by ejperezHI** [reply]
Thanks for the reply,

You were right, I missed in a "/". Now, putting the correct full path, I have a "True"
Also, the following commands give me also a "True" each of them.
The problem now (I think) is in the command `omc.sendExpression("buildModelFMU(EjemploOM)")`
That's why, after executing all this commands:
`from OMPython import OMCSessionZMQ`
`omc = OMCSessionZMQ()`
`print(omc.sendExpression("getVersion()"))`
`omc.sendExpression("loadModel(Modelica)")`
`omc.sendExpression("getErrorString()")`                  
`omc.sendExpression("loadFile(\full_path/"EjemploOM_copia.mo\")")`  
`omc.sendExpression("getErrorString()")`
`omc.sendExpression("setCommandLineOptions(\"--fmuRuntimeDepends=none\")")`
`omc.sendExpression("getErrorString()")`
`omc.sendExpression("buildModelFMU(EjemploOM)")`          
`omc.sendExpression("getErrorString()")`

The output is the same as few comments above:
![image](https://github.com/OpenModelica/OpenModelica/assets/131266350/e44b9cc4-7792-4f19-8be1-777ad4c62847)



Thanks,

**Reply by arun3688** [reply]
@ejperezHI the error message still says that the path is not loaded, please go step by step, first make sure that your path is loaded correctly, This is not the correct path 

`omc.sendExpression("loadFile(\full_path/"EjemploOM_copia.mo\")")
`

Can you put just show your local file system path of your model, because the above path looks wrong,

**Reply by ejperezHI** [reply]
Thanks agains for your reply,

Here is the full code:
`from OMPython import OMCSessionZMQ`
`omc = OMCSessionZMQ()`
`print(omc.sendExpression("getVersion()"))`
`omc.sendExpression("loadModel(Modelica)")`
`omc.sendExpression("getErrorString()")`                  
`omc.sendExpression("loadFile(\"/home/ejperez/Escritorio/OTROS_PROYECTOS/probandoOMG/EjemploOM_copia.mo\")") #This outputs me a "True"`  
`omc.sendExpression("getErrorString()") #This outputs me a ' ' just shown above`
`omc.sendExpression("setCommandLineOptions(\"--fmuRuntimeDepends=none\")")`
`omc.sendExpression("getErrorString()")`
`omc.sendExpression("buildModelFMU(EjemploOM_copia)")`          
`omc.sendExpression("getErrorString()")`

Which output is the Error image posted above.

Thanks,

**Comment by casella** [comment]
@arun3688 when we are finally through this, I think we should add this example script to the User's Guide. Please remind me about that.

**Reply by arun3688** [reply]
@casella The model is a very simple example, i guess he has some other problems with his computer which is difficult to find

**Reply by casella** [reply]
Sure, but it would be nice to have such an example anyway

**Reply by ejperezHI** [reply]
Hello,

Yes, it would be great to add an example @casella. In the other hand @arun3688 , what problem may I have? What should I look for in my computer?
I am studying it also as I post my doubts here but I still can't make the function "buildModelFMU" script work.

Just as a curiosity. Running the script "create_fmu.mos" of this [Github](https://github.com/upb-lea/OpenModelica_Microgrids/tree/master/OpenModelica_Microgrids) **I can** perfectly create FMUs. But only from default ".mo" models from the github, no from general (done by myself) .mo models. So according to this phenomenon, my computer **does can create** FMUs, but **only** through this method and only premade .mo s in that github.

Thanks a lot for the effort,
